code/go/0chain.net/smartcontract/zcnsc/stakepool.go

```diff
@@ -0,0 +1,284 @@
+		return errors.New("negative service charge")
+	}
+	if sps.MaxNumDelegates <= 0 {
+		return errors.New("num_delegates <= 0")
+	}
+
+	return nil
+}
+
+func (gn *GlobalNode) validateStakeRange(min, max state.Balance) (err error) {
+	if min < gn.MinStakeAmount {
+		return fmt.Errorf("min_stake is less than allowed by SC: %v < %v", min, gn.MinStakeAmount)
+	}
+	if max < min {
+		return fmt.Errorf("max_stake less than min_stake: %v < %v", min, max)
+	}
+
+	return
+}
+
+func (zcn *ZCNSmartContract) DistributeRewards(t *transaction.Transaction, input []byte, ctx cstate.StateContextI) (resp string, err error) {
+	return "", nil
+}
+
+func (zcn *ZCNSmartContract) AddToDelegatePool(
+	t *transaction.Transaction,
+	input []byte,
+	ctx cstate.StateContextI,
+) (resp string, err error) {
+	code := "stake_pool_lock_failed"
+
+	gn, err := GetGlobalNode(ctx)
+	if err != nil {
+		msg := fmt.Sprintf("failed to get global node, err: %v", err)
+		return "", common.NewError(code, msg)
+	}
+
+	if t.Value < gn.MinLockAmount {
+		return "", common.NewError(code, "too small stake to lock")
+	}
+
+	var spr stakePoolRequest
+	if err = spr.decode(input); err != nil {
+		return "", common.NewErrorf(code, "invalid request: %v", err)
+	}
+
+	var sp *stakePool
+	if sp, err = zcn.getStakePool(spr.AuthorizerID, ctx); err != nil {
+		return "", common.NewErrorf(code, "can't get stake pool: %v", err)
+	}
+
+	if len(sp.Pools) >= gn.MaxDelegates {
+		return "", common.NewErrorf(code, "max_delegates reached: %v, no more stake pools allowed", gn.MaxDelegates)
+	}
+
+	err = sp.LockPool(t, spenum.Authorizer, spr.AuthorizerID, spenum.Active, ctx)
+	if err != nil {
+		return "", common.NewErrorf(code, "stake pool digging error: %v", err)
+	}
+
+	if err = sp.save(zcn.ID, spr.AuthorizerID, ctx); err != nil {
+		return "", common.NewErrorf(code, "saving stake pool: %v", err)
+	}
+
+	// TO-DO: Update stake in eventDB
+
+	return
+}
+
+func (zcn *ZCNSmartContract) DeleteFromDelegatePool(t *transaction.Transaction, input []byte, ctx cstate.StateContextI) (resp string, err error) {
+	const code = "stake_pool_unlock_failed"
+	var spr stakePoolRequest
+
+	if err = spr.decode(input); err != nil {
+		return "", common.NewErrorf(code, "can't decode request: %v", err)
+	}
+	var sp *stakePool
+	if sp, err = zcn.getStakePool(spr.AuthorizerID, ctx); err != nil {
+		return "", common.NewErrorf(code, "can't get related stake pool: %v", err)
+	}
+
+	unstake, err := sp.empty(zcn.ID, spr.PoolID, t.ClientID, ctx)
+	if err != nil {
+		return "", common.NewErrorf(code,
+			"unlocking tokens: %v", err)
+	}
+
+	// the tokens can't be unlocked due to opened offers, but we mark it
+	// as 'unstake' and returns maximal time to wait to unlock the pool
+	if !unstake {
+		// save the pool and return special result
+		if err = sp.save(zcn.ID, spr.AuthorizerID, ctx); err != nil {
+			return "", common.NewErrorf(code, "saving stake pool: %v", err)
+		}
+		return toJson(&unlockResponse{Unstake: false}), nil
```

Comment:
If tokens can't be unlocked due to opened offers, the `unstake` could still return true, it would only return false when error happens, please refer the `empty()` method for details. So we might need to update the comments above, and maybe the process logic if needed

Reply:
@peterlimg Peter, I'm not following. Let's discuss in chat

Reply:
According to out talk, this is resolved